ligo/src/token/fa2.mligo

```diff
@@ -0,0 +1,155 @@
+  let valid_store = ensure_not_migrated(store) in
+  let updated_store = List.fold transfer_item params valid_store in
+  (([] : operation list), updated_store)
+
+
+// -----------------------------------------------------------------
+// Balance of
+// -----------------------------------------------------------------
+
+[@inline]
+let validate_token_type (token_id : token_id): token_id =
+  if (token_id = unfrozen_token_id || token_id = frozen_token_id) then
+    token_id
+  else (failwith ("FA2_TOKEN_UNDEFINED"): token_id)
+
+let balance_of (params, store : balance_request_params * storage): return =
+  let check_one (res_list, req : balance_response_item list * balance_request_item): balance_response_item list =
+    let valid_token_id = validate_token_type(req.token_id) in
+    match Big_map.find_opt (req.owner, valid_token_id) store.ledger with
+      Some bal ->
+        { request = req; balance = bal} :: res_list
+    | None ->
+        { request = req; balance = 0n} :: res_list
+  in
+  let result = List.fold check_one params.requests ([]: balance_response_item list) in
+  let transfer_operation = Tezos.transaction result 0mutez params.callback
+  in (([transfer_operation] : operation list), store)
+
+// -----------------------------------------------------------------
+// Token metadata registry entrypoint
+// -----------------------------------------------------------------
+
+let token_metadata_registry (params, store : token_metadata_registry_param * storage):return =
+  let cond_ = ensure_not_migrated (store) in
+  let transfer_operation = Tezos.transaction store.token_address 0mutez params
+  in
+    (([transfer_operation] : operation list), store)
+
+
+// -----------------------------------------------------------------
+// Update operators entrypoint
+// -----------------------------------------------------------------
+[@inline]
+let validate_operator_token (token_id : token_id): token_id =
+  if (token_id = unfrozen_token_id) then
+    token_id
+  else if (token_id = frozen_token_id) then
+    (failwith ("OPERATION_PROHIBITED"): token_id)
+  else
+    (failwith ("FA2_TOKEN_UNDEFINED"): token_id)
+
+let update_one (store, param: storage * update_operator): storage =
+  let (operator_update, operator_param) =
+    match param with
+      Add_operator p -> (Some unit, p)
+    | Remove_operator p -> ((None : unit option), p)
+  in
+  let valid_token_id = validate_operator_token (operator_param.token_id) in
+  if (sender = operator_param.owner) then
+    let key: operator = { owner = operator_param.owner; operator = operator_param.operator} in
+    let updated_operators = Big_map.update key operator_update store.operators
+    in  { store with
+          operators = updated_operators
+        }
+  else (failwith("NOT_OWNER"): storage)
+
+let update_operators (params, store : update_operators_param * storage):return =
+  let cond_ = ensure_not_migrated (store) in
+  let updated_store = List.fold update_one params store in
```

Comment:
```suggestion
  let valid_store = ensure_not_migrated (store) in
  let updated_store = List.fold update_one params valid_store in
```

Reply:
Btw, this practice of calling store variables different each time does not seem good to me.
1. I have to invent a new name for storage variable each time I call a function.
2. Now I have to be **extra** careful when updating the code, because it is too easy to use an outdated variable, and LIGO won't warn me if the most recent storage variable remained unused.

In proposals logic I have methods which perform up to 3 checks and then 2-3 updates, and with regard to this practice those issues seem quite severe.

Maybe we could just call the variable as `store`  everywhere? LIGO seems not to care about a variable with the same name being defined multiple times, and particularly for storage variable this seems to be much safer. Though I don't know whether will it work as expected.

At the very least maybe use `store1`, `store2` e.t.c.? This way (1) is not a problem anymore, and issues with (2) can be relatively easy detected.

Reply:
I agree that making verification return `valid_store` and let further functions work with validated storage is generally a good practice, but to me in makes sense only in a language where `store` and `valid_store` would have different types, while here the compiler does not prevent us from abusing this pattern accidentially. After all, we only gain many different ways to implement the same thing (pick/ignore validated store, how to call the returned variables, how to chain several checks, e.t.c) and more points for mutual nitpicking :smile:

Reply:
@Martoon-00 you make good points, I though about those too yesterday and didn't come up with a way to improve this substantially.

Regarding checks the good think is that these can often be made all at once and have for example something like:
```haskell
  let valid_store = check1 (check2 (check3 store)) in
```

I also tried to make a function composition, but since this isn't actually `OCaml` it seems to be only doable with currying which impacts on produced Michelson, otherwise it would have looked like this:
```haskell
  let updated_store =
    store
    |> ensure_not_migrated
    |> List.fold update_one params
```

I think it's fine to call the variable `store` anywhere since replacement seems to be working fine, but we should be extra careful this does not produce wrong code (tests should catch that tho).

@rinn7e if you agree as well could you make these changes?

Reply:
Yeah, pretty sad that LIGO is limited here, without all the conveniences we used to in Haskell it is actually difficult to write on a language with immutable variables.

Reply:
Yeah, I'm good with just calling the it `store` everywhere as well.

Reply:
I've pushed the changes now, let me know if this is good enough.

Reply:
Thanks!